libcnb/src/build.rs

```diff
@@ -95,18 +104,297 @@ impl<B: Buildpack + ?Sized> BuildContext<B> {
+    ///         layer_ref.replace_env(LayerEnv::new().chainable_insert(
+    ///             Scope::All,
+    ///             ModificationBehavior::Append,
+    ///             "PLANET",
+    ///             "LV-246",
+    ///         ))?;
+    ///     }
+    ///     LayerState::Restored { .. } => {
+    ///         println!("Reusing example layer from previous run!");
+    ///     }
+    /// }
+    /// #
+    /// #        BuildResultBuilder::new().build()
+    /// #    }
+    /// # }
+    /// #
+    /// # impl From<ExampleBuildpackError> for libcnb::Error<ExampleBuildpackError> {
+    /// #    fn from(value: ExampleBuildpackError) -> Self {
+    /// #        Self::BuildpackError(value)
+    /// #    }
+    /// # }
+    /// ```
+    ///
+    /// # More complex example
+    /// ```rust
+    /// # use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+    /// # use libcnb::detect::{DetectContext, DetectResult};
+    /// # use libcnb::generic::GenericPlatform;
+    /// # use libcnb::layer::{
+    /// #     CachedLayerDefinition, EmptyLayerCause, InspectRestoredAction, InvalidMetadataAction,
+    /// #     LayerState,
+    /// # };
+    /// # use libcnb::Buildpack;
+    /// # use libcnb_data::generic::GenericMetadata;
+    /// # use libcnb_data::layer_name;
+    /// # use serde::{Deserialize, Serialize};
+    /// # use std::fs;
+    /// #
+    /// # struct ExampleBuildpack;
+    /// #
+    /// # #[derive(Debug)]
+    /// # enum ExampleBuildpackError {
+    /// #     UnexpectedIoError(std::io::Error),
+    /// # }
+    /// #
+    /// #[derive(Deserialize, Serialize)]
+    /// struct ExampleLayerMetadata {
+    ///     lang_runtime_version: String,
+    /// }
+    ///
+    /// enum CustomCause {
+    ///     Ok,
+    ///     LegacyVersion,
+    ///     HasBrokenModule,
+    ///     MissingModulesFile,
+    /// }
+    ///
+    /// # impl Buildpack for ExampleBuildpack {
+    /// #     type Platform = GenericPlatform;
+    /// #     type Metadata = GenericMetadata;
+    /// #     type Error = ExampleBuildpackError;
+    /// #
+    /// #     fn detect(&self, _: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+    /// #         unimplemented!()
+    /// #     }
+    /// #
+    /// fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+    ///     let layer_ref = context.cached_layer(
+    ///         layer_name!("example_layer"),
+    ///         CachedLayerDefinition {
+    ///             build: false,
+    ///             launch: false,
+    ///             invalid_metadata: &|_| InvalidMetadataAction::DeleteLayer,
+    ///             inspect_restored: &|metadata: &ExampleLayerMetadata, layer_dir| {
+    ///                 if metadata.lang_runtime_version.starts_with("0.") {
+    ///                     Ok((
+    ///                         InspectRestoredAction::DeleteLayer,
+    ///                         CustomCause::LegacyVersion,
+    ///                     ))
```

Comment:
I found the tuple being returned here confusing, since I was expecting to see a `From` / `Into` implementation in the rustdocs here (based on the last time I'd seen the proposed design) but there wasn't one.

As I continued the review, I found the built-in `From` implementations for the common forms (such as returning a tuple) - however, these don't appear to be documented in the rustdocs anywhere.

I'm not sure where is best to put these? Perhaps the `CachedLayerDefinition` rustdocs, and then link to those from the rustdocs of `cached_layer` with a see more at X? Or vice versa.

Reply:
Added some additional docs in [400b1b3](https://github.com/heroku/libcnb.rs/pull/814/commits/400b1b334181e5d2036a4aa13a586a464ce38ae9)